src/app/api/game-sessions/route.ts

```diff
@@ -1,16 +1,87 @@
 import type { NextRequest } from "next/server";
 import { NextResponse } from "next/server";
+import { and, eq, gt, gte, lt, lte, or } from "drizzle-orm";
 import z from "zod";
 
 import { db } from "@/lib/db";
-import { gameSessions } from "@/lib/db/schema";
-import { insertGameSessionSchema } from "@/lib/validators";
+import {
+  gameSessionExceptions,
+  gameSessions,
+  gameSessionSchedules,
+  semesters,
+} from "@/lib/db/schema";
+import { getCurrentUser } from "@/lib/session";
+import { getWeekday } from "@/lib/utils";
+import {
+  insertGameSessionExceptionSchema,
+  insertGameSessionSchema,
+  updateGameSessionSchema,
+} from "@/lib/validators";
 
-export async function GET() {
+export async function GET(req: NextRequest) {
   try {
-    const sessions = await db.query.gameSessions.findMany();
-    return NextResponse.json(sessions);
-  } catch {
+    const user = await getCurrentUser();
+    if (!user) {
+      return new Response("ERROR: Unauthorized request", { status: 401 });
+    }
+    if (user.role != "admin") {
+      return new Response("ERROR: No valid permissions", { status: 403 });
+    }
+
+    const date = req.nextUrl.searchParams.get("date");
+    const gameSessionDate = z.string().date().parse(date);
+
+    const gameSession = await db.query.gameSessions.findFirst({
+      where: eq(gameSessions.date, gameSessionDate),
+    });
+
+    if (gameSession) {
+      // If a game session exists for the date, return it
+      return NextResponse.json(gameSession, { status: 200 });
+    }
+
+    // Check for a schedule on this weekday
+    const semester = await db.query.semesters.findFirst({
+      with: {
+        gameSessionSchedules: {
+          where: eq(gameSessionSchedules.weekday, getWeekday(gameSessionDate)),
+        },
+      },
+      where: or(
+        and(
+          lte(semesters.startDate, gameSessionDate),
+          gt(semesters.breakStart, gameSessionDate),
+        ),
+        and(
+          lt(semesters.breakEnd, gameSessionDate),
+          gte(semesters.endDate, gameSessionDate),
+        ),
+      ),
+    });
+
+    const gameSessionSchedule = semester?.gameSessionSchedules[0];
+
+    if (!gameSessionSchedule) {
+      // If no game session schedule found for the weekday within the semester, return
+      return new Response(null, { status: 204 });
+    }
+
+    const gameSessionException = await db.query.gameSessionExceptions.findFirst(
+      {
+        where: eq(gameSessionExceptions.gameSessionDate, gameSessionDate),
+      },
+    );
+
+    if (gameSessionSchedule && !gameSessionException) {
+      // If schedule found and no exception, return the schedule
+      return NextResponse.json(gameSessionSchedule, { status: 200 });
+    } else {
+      return new Response(null, { status: 204 });
+    }
```

Comment:
nit: if you have a return statement in the if block then you don't really need an else block.